lib/AST/ASTDumper.cpp

```diff
@@ -3593,6 +3701,538 @@ class PrintTypeRepr : public TypeReprVisitor<PrintTypeRepr, void, StringRef>,
     printFoot();
   }
 };
+} // end anonymous namespace
+
+//===----------------------------------------------------------------------===//
+// Dumping for DeclAttributes
+//===----------------------------------------------------------------------===//
+
+namespace {
+class PrintAttribute : public AttributeVisitor<PrintAttribute, void, StringRef>,
+                       public PrintBase {
+  const ASTContext *Ctx;
+
+public:
+  PrintAttribute(
+      raw_ostream &os, const ASTContext *ctx, unsigned indent = 0,
+      bool parseIfNeeded = false,
+      llvm::function_ref<Type(Expr *)> getTypeOfExpr = defaultGetTypeOfExpr,
+      llvm::function_ref<Type(TypeRepr *)> getTypeOfTypeRepr = nullptr,
+      llvm::function_ref<Type(KeyPathExpr *E, unsigned index)>
+          getTypeOfKeyPathComponent = defaultGetTypeOfKeyPathComponent)
+      : PrintBase(os, indent, parseIfNeeded, getTypeOfExpr, getTypeOfTypeRepr,
+                  getTypeOfKeyPathComponent),
+        Ctx(ctx) {}
+
+  void printCommon(DeclAttribute *Attr, StringRef name, StringRef label) {
+    printHead(name, DeclAttributeColor, label);
+    printFlag(Attr->isImplicit(), "implicit");
+    printFlag(Attr->isInvalid(), "invalid");
+    printFlag(Attr->getAddedByAccessNote(), "added_by_access_note");
+    printSourceRange(Attr->Range, Ctx);
+  }
+
+  /// Deleting this ensures that all attributes are covered by the
+  /// visitor below.
+  void visitDeclAttribute(DeclAttribute *A) = delete;
+
+#define TRIVIAL_ATTR_PRINTER(Class, Name)                                      \
+  void visit##Class##Attr(Class##Attr *Attr, StringRef label) {                \
+    printCommon(Attr, #Name "_attr", label);                                   \
+    printFoot();                                                               \
+  }
+
+  TRIVIAL_ATTR_PRINTER(Actor, actor)
```

Comment:
Any reason not to use `SIMPLE_DECL_ATTR` here? LLVM has `convertToSnakeFromCamelCase` FWIW which could be used to convert the class name

Reply:
Oooh, I didn't know `convertToSnakeFromCamelCase`. ... it seems to convert `ObjCMembers` to `obj_c_members`? which makes sense, but since we use `objc_`, I'd go without it for now. Unfortunate.